---
title: "nanoNOMe bulk repeat methylation analysis"
author: Isac Lee
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_notebook : default
  pdf_document:
    citation_package: natbib
    keep_tex: true
    fig_caption: true
    latex_engine: xelatex
#bibliography: master.bib
# set dir in the following function to the desired loc
mainfont: "DejaVu Sans" # Try out some font options if xelatex
titlefont: "DejaVu Sans" # Try out some font options if xelatex
---

```{r setup, eval=TRUE, include=FALSE, cache=F, message=F, warning=F, results="hide"}
rm(list=ls());gc()
knitr::opts_chunk$set(fig.path='figs/')
knitr::opts_chunk$set(cache = FALSE, warning = FALSE, 
                      message = FALSE, cache.lazy = FALSE)
my_plot_hook <- function(x, options)
  paste("\n", knitr::hook_plot_tex(x, options), "\n")
knitr::knit_hooks$set(plot = my_plot_hook)
```

```{r libs, eval=T, include=FALSE, cache=F, message=F, warning=F, results="hide"}
source("~/Code/ilee/plot/ilee_plot_utils.R")
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
library(bsseq)
library(biomaRt)
library(UpSetR)
library(grid)
library(ggridges)
```

```{r functions, include = F}
getQQ <- function(one,two,qs = seq(0,1,0.01)) {
 tibble(
   one = quantile(one,qs, na.rm = T),
   two = quantile(two,qs, na.rm = T), 
   qunatile = qs)
}
getRegMeth <- function(bs.list,reg){
  # get region methylation into my favorite format
  bs.reg <- lapply(bs.list,function(x)x[overlapsAny(x,reg),])
  # coverage
  keepi.list <- lapply(bs.reg,function(x){
    cov <- getCoverage(x,type="Cov",what="perBase")
    rowSums(cov > 3) == ncol(x)
  })
  bs.reg <- lapply(seq_along(bs.reg),function(i){
    bs.reg[[i]][keepi.list[[i]],]
  })
  names(bs.reg) <- names(bs.list)
  ## get methylation
  meth.list <- lapply(bs.reg,function(x){
    m <- as_tibble(getMeth(x,type = "smooth",what = "perBase"))
    colnames(m) <- pData(x)$allele
    m %>% 
      bind_cols(as_tibble(granges(x)) %>% 
                  dplyr::select(seqnames,start,end))
  })
  
  # merge
  meth <- do.call(rbind,list(
    meth.list$cpg %>% mutate(mod = "CpG Methylation"),
    meth.list$gpc %>% mutate(mod = "GpC Accessibility")
  ))
  meth.gr <- GRanges(meth[,c("seqnames","start","end")])
  
  # gather 
  meth.gather <- meth %>%
    gather(hap,freq,-seqnames,-start,-end,-mod)
  meth.gather
    
}
```

```{r annotation, eval = T, include = FALSE, cache=F }
################################################
# genes
################################################

mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
trans <- getBM(attributes = c("chromosome_name", 
                              "start_position",
                              "end_position",
                              "strand",
                              "entrezgene_id",
                              "ensembl_gene_id",
                              "ensembl_transcript_id",
                              "hgnc_symbol"), mart = mart)
trans.all <- as_tibble(trans) %>%
  dplyr::rename(chrom = chromosome_name,start = start_position, end = end_position) %>%
  arrange(chrom,start,end) %>%
  mutate(chrom = 
           ifelse(grepl("chr",chrom),
                  chrom,paste0("chr",chrom)),
         strand = ifelse(strand == 1,"+","-")) 
trans <- trans.all %>%
  distinct(chrom,start,end,strand,.keep_all=T)
regs <- resize(GRanges(trans),width = trans$end-trans$start + 4000,fix = "center")
# filter duplicate TSS's
tss.all <- promoters(GRanges(trans),upstream=1,downstream=0)
keepi <- !duplicated(tss.all)
tss <- tss.all[keepi]

dir <- "/kyber/Data/Nanopore/projects/nanonome/analysis/data/gm12878"
################################################
# H3K4me3
################################################
h3k4me3_path <- file.path(dir,"GM12878_H3K4me3_ChIP_peaks.bed.gz")
h3k4me3.peaks <- GRanges(read_tsv(h3k4me3_path,col_names = c("chrom","start","end")) )
h3k27me3_path <- file.path(dir,"GM12878_H3K27me3_ChIP_peaks.bed.gz")
h3k27me3.peaks <- GRanges(read_tsv(h3k27me3_path,col_names = c("chrom","start","end")) )
proms <- resize(tss,width = 2e3,fix = "center")
k4ovl <- findOverlaps(proms,h3k4me3.peaks)
k27ovl <- findOverlaps(proms,h3k27me3.peaks)
tss$H3K4me3 <- tss$H3K27me3 <- "None"
tss$H3K4me3[queryHits(k4ovl)] <- "H3K4me3" 
tss$H3K27me3[queryHits(k27ovl)] <- "H3K27me3" 
tss$histone <- case_when(
    tss$H3K4me3 == "H3K4me3" & tss$H3K27me3 == "H3K27me3" ~ "both",
    tss$H3K4me3 == "H3K4me3" ~ "H3K4me3",
    tss$H3K27me3 == "H3K27me3" ~ "H3K27me3",
    TRUE ~ "None"
  )
```

```{r repeats, eval = T, include = FALSE, cache=T }
# first all repetitive elements?
rep_fh <- "/kyber/Data/Nanopore/projects/nanonome/analysis/data/hg38/hg38_repeats.bed"
rep_cnames <- c("chrom","start","end","repname","repscore", "strand","reptype")
reps <- read_tsv(rep_fh,col_names = rep_cnames) 
# fix start to be 1-based, add width, and add label
reps <- reps %>%
  mutate(
    start = start - 1,
    width = end - start + 1,
    label = paste(chrom,start,end,repname,sep ="_"))
# filter some rep type
reps <- reps %>%
  mutate(reptype = case_when(
    grepl("lu",repname) ~ "Alu",
    grepl("MIR",repname) ~ "MIR",
    TRUE ~ reptype ))
reps <- reps %>%
  filter(reptype %in% c("LINE","Alu","MIR","LTR","MIR"))
table(reps$reptype)
reps.gr <- GRanges(reps)
```

#### data

```{r data , eval = T, include = FALSE, cache=F }
plotdir <- "/home/isac/Dropbox (Timp Lab)/Data/nome-seq/version3_guppy3/plots/frequency"
# data
fdir = "/uru/Data/Nanopore/projects/nanonome/pooled/mfreq"
cpgpath <- file.path(fdir,"GM12878_nanoNOMe.cpg.BSmooth.rds")
gpcpath <- file.path(fdir,"GM12878_nanoNOMe.gpc.BSmooth.rds")
peakpath <- "/uru/Data/Nanopore/projects/nanonome/peaks/GM12878_nanoNOMe_accessibility_peaks.rds"

# methylation data
cpg <- readRDS(cpgpath)
gpc <- readRDS(gpcpath)
bs.list <- list(cpg,gpc)
names(bs.list) <- c("cpg","gpc")
# only use coordinates that have > 5x coverage
cov.list <- lapply(bs.list,getCoverage,type = "Cov",what = "perBase")
keepi.list <- lapply(cov.list,function(x)which(x > 5))
bs.list <- mapply(function(x,i)x[i,],bs.list,keepi.list)

# peaks
peaks <- readRDS(peakpath)
# let's just use significant peaks
peaks <- peaks %>% filter(sig == "sig")
peaks.gr <- GRanges(peaks)
```

### filter and randomize regions based on coverage in data

1. Get clusters with default widths to be able to get total data lengths for each region  
2. Get clusters with max 10e3 gap to get regions with data
3. subset repeats that overlap with (2)
4. randomize repeat coordinates in regions in (2), and subset those that do not overlap with (3)

```{r regions, eval = T, include = FALSE, cache=T }
# data-containing clusters
cid.list <- lapply(bs.list,function(x){
  bsseq:::clusterMaker(as.character(seqnames(x)),start(x))})
clusters.list <- mapply(function(x,i){
  chrs <- as.character(seqnames(x))
  positions <- start(x)
  tibble(idx = seq_along(i),cid = i) %>%
    group_by(cid) %>%
    summarize(idxStart = min(idx), idxEnd= max(idx)) %>%
    mutate(chr = chrs[idxStart], 
           start = positions[idxStart],
           end = positions[idxEnd])
},bs.list,cid.list, SIMPLIFY = F)
# which regions are covered in data - max gap of 10kb
cid.list <- lapply(bs.list,function(x){
  bsseq:::clusterMaker(as.character(seqnames(x)),start(x),maxGap = 10e3)})
datregs.list <- mapply(function(x,i){
  chrs <- as.character(seqnames(x))
  positions <- start(x)
  tibble(idx = seq_along(i),cid = i) %>%
    group_by(cid) %>%
    summarize(idxStart = min(idx), idxEnd= max(idx)) %>%

        mutate(chr = chrs[idxStart], 
           start = positions[idxStart],
           end = positions[idxEnd])
},bs.list,cid.list, SIMPLIFY = F)
# only use regions that have data in both mod
regs.gr <- intersect(GRanges(datregs.list$cpg), GRanges(datregs.list$gpc))
# let's remove chrM and chrY
regs.gr <- regs.gr[!seqnames(regs.gr) %in% c("chrY","chrM")]

# reps in regs - type = "within" ensures that the total length of repeats are covered in regions
keepi <- overlapsAny(reps.gr,regs.gr,type = "within")
table(keepi)
regreps <- reps[keepi,]
regreps.gr <- reps.gr[keepi]

# random coordinates for reference
chromsizes <- read_tsv("/mithril/Data/NGS/Reference/hg38_noalt/hg38_noalt.fa.fai",col_names = c("chrom","size"))

shuf <- regreps %>%
  rowwise() %>%
  mutate( 
    start = floor(runif(1,min = 1, max = chromsizes$size[match(chrom,chromsizes$chrom)] - width)), 
    end = start + width - 1, 
    strand = strand) %>%
  as_tibble()
shuf.gr <- GRanges(shuf)
# remove shuf not in regs
keepi <- overlapsAny(shuf.gr,regs.gr,type = "within")
table(keepi)
shuf <- shuf[keepi,]
shuf.gr <- shuf.gr[keepi]
# remove shuf overlapping rep
keepi <- !overlapsAny(shuf.gr,regreps.gr)
table(keepi)
shuf <- shuf[keepi,]
shuf.gr <- shuf.gr[keepi]

table(shuf$reptype)
table(regreps$reptype)
# combine regs
allregs <- bind_rows(list(rep = regreps,shuf = shuf),.id = "what")
allregs.gr <- GRanges(allregs)
```

### get region methylation 

```{r cpg_meth, eval = T, include = FALSE, cache=F }
# reps - cpg
cpg_m <- getCoverage(bs.list$cpg,region = allregs.gr,type = "M", what = "perRegionTotal")
cpg_cov <- getCoverage(bs.list$cpg,region = allregs.gr,type = "Cov", what = "perRegionTotal")
cpg.tb <- allregs %>%
  mutate(meth = cpg_m[,1], cov = cpg_cov[,1])
# gpc - how about peak accessibility?
gpc_m <- getMeth(bs.list$gpc,region = allregs.gr, type = "smooth", what = "perBase")
gpc_cov <- getCoverage(bs.list$gpc,region = allregs.gr, type = "Cov", what = "perBase")
gpc_m_max <- mapply(function(m,cov){
  keepi <- cov > 20
  max(m[keepi],na.rm = T)
  }, gpc_m,gpc_cov)
# peak accessibility 
meth.tb <- cpg.tb %>%
  mutate(gpc = gpc_m_max,
         cpg = meth/cov) %>%
  filter(cov > 20) %>%
  mutate(gpc = ifelse(gpc == -Inf, NA, gpc))

# gather
meth.gather <- meth.tb %>%
  gather(mod,freq,cpg,gpc)

# gpc peaks
peaki <- overlapsAny(allregs.gr,peaks.gr)
peaks.tb <- cpg.tb %>%
  mutate(peak = peaki) %>%
  filter(cov > 20)
peaks.frac <- peaks.tb %>%
  group_by(peak,what,reptype) %>%
  summarize(n =n ()) %>%
  spread(peak,n) %>%
  mutate(frac = `TRUE`/(`TRUE`+`FALSE`))
peaks.sum <- peaks.frac %>%
  dplyr::select(what,reptype,frac) %>%
  spread(what,frac) %>%
  mutate(ratio = rep/shuf)
# peaks per 1mb and enrichment?
gpc.cl <- clusters.list$gpc
gpc.size <- sum(gpc.cl$end - gpc.cl$start + 1)
global <- length(peaks.gr)/gpc.size * 1e6
ovl <- findOverlaps(peaks.gr,regreps.gr)
repn <- regreps[subjectHits(ovl),] %>%
  mutate(peakidx = queryHits(ovl)) %>%
  group_by(reptype) %>%
  summarize( n = n())
peaknums <- regreps %>%
  group_by(reptype) %>%
  summarize( width = sum(width)) %>%
  mutate(n = repn$n[match(reptype,repn$reptype)],
    normn = n / width * 1e6
  )  %>%
  bind_rows(tibble(reptype = "Global",normn = global))
```

```{r meth_plot, eval = T, echo = F, fig.height=3, fig.width=6, message=F, warning = F}
# rename what
meth.plt <- meth.gather %>%
  mutate(Region = ifelse(what == "rep","Observed","Expected")) 
plotdir <- "/home/isac/Dropbox/Data/nome-seq/version3_guppy3/plots/frequency"
plotpath <- file.path(plotdir,"200218_gm12878_repeats_density.pdf")
pdf(plotpath,height = 3, width = 6, useDingbats = F)
ggplot(meth.plt,aes( x = freq, y = reptype, alpha = Region, linetype = Region)) +
  facet_wrap(~mod, ncol = 2,scales = "free", labeller = as_labeller(c(cpg="Region CpG Methylation",gpc="Peak GpC Accessibility")), nrow = 2) +
  scale_linetype_manual(values = c("dashed","solid")) +
  scale_alpha_manual(values = c(0,0.5)) +
  geom_density_ridges(scale = 1.5, fill = "orange") +
  labs( x = "Frequency", y = "Repeat Type") +
  theme(legend.background = element_rect(fill = NA, color = "black", linetype = "solid")) +
  coord_cartesian(clip = "off")
dev.off()
```
```{r peak_frac_plot, eval = T, echo = F, fig.height=1.5, fig.width=3, message=F, warning = F}
peaks.plt <- peaks.frac %>%
  mutate(Region = ifelse(what == "rep","Repeat","Shuffled"))
peaks.enrich <- peaks.plt %>%
  ungroup() %>%
  dplyr::select(reptype,frac,Region) %>%
  spread(Region,frac)  %>%
  mutate(enrich = Repeat/Shuffled)
plotpath <- file.path(plotdir,"200218_gm12878_repeats_peak_fractions.pdf")
pdf(plotpath,height = 3, width = 2, useDingbats = F)
ggplot(peaks.plt,aes( x = reptype, fill = Region, y = frac)) +
  geom_histogram(stat = "identity",position = "dodge") +
  labs( x = "Repeat Type", y = "Fraction with\nAccessibility Peak") +
  theme(legend.background = element_rect(fill = NA, color = "black", linetype = "solid")) +
  coord_cartesian(clip = "off")
ggplot(peaks.enrich,aes( x = reptype, y = enrich)) +
  geom_histogram(stat = "identity",position = "dodge") +
  geom_hline(yintercept = 1, linetype = "dashed") +
  labs( x = "Repeat Type", y = "Fraction with\nAccessibility Peak") +
  theme(legend.background = element_rect(fill = NA, color = "black", linetype = "solid")) +
  coord_cartesian(clip = "off")
plt <- peaknums %>%
  mutate(reptype = factor(reptype,levels = c("Global","Alu","LINE","LTR","MIR")),
    nlab = as.character(round(normn))
  )
ggplot(plt,aes( x = reptype, y = normn)) +
  geom_histogram(stat = "identity",position = "dodge") +
  labs( x = "Repeat Type", y = "Number of\nPeaks Per 1Mb") +
  theme(legend.background = element_rect(fill = NA, color = "black", linetype = "solid")) +
  geom_text(mapping = aes(label = nlab), hjust = 0) +
  coord_flip(clip = "off") 
#  coord_cartesian(clip = "off")
dev.off()
```
### per cpg methylation

```{r}
cpg_m <- getMeth(bs.list$cpg,region = allregs.gr,type = "raw", what = "perBase")
allregs
m.list <- lapply(seq_along(cpg_m),function(i){
  x <- cpg_m[[i]][,1]
  if (!is.null(x)){
    y <- tibble(freq = x)
    bind_cols(y,allregs[rep(i,nrow(y)),])
  }
  })

m.tb <- bind_rows(m.list)
table(m.tb$reptype)
plotpath <- file.path(plotdir,"200426_gm12878_repeat_per_site_cpg.pdf")
pdf(plotpath,height = 4, width = 5, useDingbats = F)
ggplot(m.tb,aes( x = freq, y = reptype, alpha = what, linetype = what)) +
  scale_linetype_manual(values = c("dashed","solid")) +
  scale_alpha_manual(values = c(0,0.5)) +
  geom_density_ridges(scale = 1.5, fill = "orange") +
  labs( x = "Frequency", y = "Repeat Type") +
  theme(legend.background = element_rect(fill = NA, color = "black", linetype = "solid")) +
  coord_cartesian(clip = "off")
dev.off()
#ggplot(m.tb,aes(x = freq, color = what)) +
#  facet_wrap(~reptype) +
#  geom_density()


```

### methylation vs peaks

Methylation at repeats with accessibility peaks vs not?

```{r peak_meth, eval = T, echo = F, fig.height=3, fig.width=4, message=F, warning = F}
# per base methylation at regions
cpg.rep <- bs.list$cpg[overlapsAny(regreps.gr),]
ovl <- findOverlaps(cpg.rep,regreps.gr)
cpgmeth.rep <- getMeth(cpg.rep,what = "perBase")
reps.meth <- regreps[subjectHits(ovl),] %>%
  mutate(
    pos = start(cpg.rep[queryHits(ovl)]),
    cpg = cpgmeth.rep[queryHits(ovl),1])
repsmeth.gr <- GRanges(reps.meth %>% mutate( start = pos, end = pos))
peaki <- overlapsAny(repsmeth.gr,peaks.gr)
reps.meth <- reps.meth %>%
  mutate(peak = ifelse(peaki, "With Peak","No Peak")) %>%
  group_by(chrom,start,end,repname) %>%
  mutate(regpeak = ifelse("With Peak" %in% peak, "With Peak","No Peak")) %>%
  ungroup()
repregs.meth <- reps.meth %>%
  group_by(chrom,start,end,repname,reptype,regpeak) %>%
  summarize( cpg = mean(cpg))
nums <- repregs.meth %>%
  group_by(reptype,regpeak) %>%
  summarize( n =n())
ggplot(repregs.meth %>% mutate(width = end-start + 1),aes( x = width, y = reptype, fill = regpeak)) +
  geom_density_ridges(alpha = 0.5,scale = 1.5) +
  lims(x= c(0,5000))

ggplot(repregs.meth,aes( x = cpg,fill = regpeak, y = reptype)) +
  geom_density_ridges(alpha = 0.5,scale = 1.5) +
  labs( x = "Frequency", y = "Repeat Type") +
  theme(legend.background = element_rect(fill = NA, color = "black", linetype = "solid")) +
  coord_cartesian(clip = "off")
```
## Only Alus

Because other repeat types don't suggest methylation plays a big role, let's just examine Alu repeats : 

```{r alu_meth, eval = T, echo = F, fig.height=3, fig.width=4, message=F, warning = F}
aluregs.meth <- repregs.meth %>%
  filter(reptype == "Alu") %>%
  mutate(regpeak = ifelse(regpeak == "No Peak","No Accessibility Peak","Accessibility Peak within Repeat"))
ggplot(aluregs.meth,aes( x = cpg,fill = regpeak)) +
  facet_wrap(~regpeak,nrow = 2,scales = "free") +
  geom_histogram(alpha = 0.5,scale = 1.5) +
  labs( x = "Frequency", y = "Repeat Type") +
  theme(legend.position = "none") +
  coord_cartesian(clip = "off")
```

### Genomic context of Alus with peaks

First histone binding/ TSS

```{r alu_histone, eval = T, echo = F, fig.height=3, fig.width=4, message=F, warning = F}
aluregs.gr <- GRanges(aluregs.meth)
tssdists <- as.data.frame(distanceToNearest(aluregs.gr,tss, ignore.strand = T))
k4dists <- as.data.frame(distanceToNearest(aluregs.gr,h3k4me3.peaks, ignore.strand = T))
k27dists <- as.data.frame(distanceToNearest(aluregs.gr,h3k27me3.peaks, ignore.strand = T))

# how often are accessibility peaks overlapping k4/k27 peaks?
# k4 overlaps about half the time with peaks
peaks_k3 <- peaks.gr[overlapsAny(peaks.gr,h3k4me3.peaks)]
k3tsspeaks.tb <- as.data.frame(distanceToNearest(peaks_k3,tss,ignore.strand = T))
summary(k3tsspeaks.tb$distance)
# let's do instead a global comparison of peak enrichmnet in various contexts and methylation in them?
# k27 barely overlaps with peaks
#k27_peaks <- overlapsAny(peaks.gr,h3k27me3.peaks)
#table(k27_peaks)
peaks.gr
aluregs.meth <- aluregs.meth %>%
  ungroup() %>%
  mutate(tssdist = tssdists$distance,
         k4dist = k4dists$distance,
         tssidx = tssdists$subjectHits,
         k4 = ifelse(k4dist == 0, "K4","noK4"),
         histone = tss$histone[tssidx])
aluregs.meth %>%
  group_by(regpeak) %>%
  summarize( n = n(), meandist = mean(k4dist), maxdist = max(k4dist))
aluregs.meth %>% 
  group_by(regpeak,k4) %>%
  summarize( n = n())
  
ggplot(aluregs.meth %>% filter(k4 == "noK4", histone != "H3K4me3"),aes( x = cpg)) +
  facet_wrap(~regpeak,scales = "free",nrow = 2) +
  geom_histogram() +
  theme(legend.position = "bottom")

```
```{r alu_other, eval = T, echo = F, fig.height=3, fig.width=4, message=F, warning = F}
# use unmappable data for coordinates of other annotated regions
unmap_fh <- "/kyber/Data/Nanopore/projects/nanonome/analysis/mappability/200218_wgbs_unmappable_regions.rds"
unmapregs <- readRDS(unmap_fh) %>%
  ungroup()  %>%
  mutate(what = ifelse(nowgbs == T, "NoWGBS","YesWGBS"))
table(unmapregs$type)
 
```

### Regions unmappable by WGBS

```{r unmappable, eval = T, echo = F, fig.height=1.5, fig.width=3, message=F, warning = F}
unmap_fh <- "/kyber/Data/Nanopore/projects/nanonome/analysis/mappability/200218_wgbs_unmappable_regions.rds"
unmapregs <- readRDS(unmap_fh) %>%
  ungroup()  %>%
  mutate(what = ifelse(nowgbs == T, "NoWGBS","YesWGBS"))
 
unmapregs.gr <- GRanges(unmapregs) 
```

### get region methylation 


```{r unmap_meth, eval = T, include = FALSE, cache=F }
# reps - cpg
cpg_m <- getCoverage(bs.list$cpg,region = unmapregs.gr,type = "M", what = "perRegionTotal")
cpg_cov <- getCoverage(bs.list$cpg,region = unmapregs.gr,type = "Cov", what = "perRegionTotal")
cpg.tb <- unmapregs %>%
  mutate(meth = cpg_m[,1], cov = cpg_cov[,1])
# gpc - how about peak accessibility?
gpc_m <- getMeth(bs.list$gpc,region = unmapregs.gr, type = "smooth", what = "perBase")
gpc_cov <- getCoverage(bs.list$gpc,region = unmapregs.gr, type = "Cov", what = "perBase")
gpc_m_max <- mapply(function(m,cov){
  keepi <- cov > 20
  max(m[keepi],na.rm = T)
  }, gpc_m,gpc_cov)
# peak accessibility 
meth.tb <- cpg.tb %>%
  mutate(gpc = gpc_m_max,
         cpg = meth/cov) %>%
  filter(cov > 20) %>%
  mutate(gpc = ifelse(gpc == -Inf, NA, gpc))

# gather
meth.gather <- meth.tb %>%
  gather(mod,freq,cpg,gpc)
summary(meth.gather$freq)

# gpc peaks
peaki <- overlapsAny(unmapregs.gr,peaks.gr)
peaks.tb <- cpg.tb %>%
  mutate(peak = peaki) %>%
  filter(cov > 20)
peaks.frac <- peaks.tb %>%
  group_by(peak,what,type) %>%
  summarize(n =n ()) %>%
  spread(peak,n) %>%
  mutate(frac = `TRUE`/(`TRUE`+`FALSE`))
peaks.sum <- peaks.frac %>%
  dplyr::select(what,type,frac) %>%
  spread(what,frac) %>%
  mutate(ratio = NoWGBS/YesWGBS)
```

```{r unmap_meth_plot, eval = T, echo = F, fig.height=3, fig.width=6, message=F, warning = F}
# rename what
meth.plt <- meth.gather %>%
  mutate(Region = ifelse(what == "NoWGBS","Unmapped","Mapped"))
plotpath <- file.path(plotdir,"200218_gm12878_unmapped_density.pdf")
#pdf(plotpath,height =3, width = 6, useDingbats = F)
ggplot(meth.plt,aes( x = freq, fill = Region, y = type)) +
  facet_wrap(~mod, ncol = 2,scales = "free", labeller = as_labeller(c(cpg="Region CpG Methylation",gpc="Peak GpC Accessibility")), nrow = 2) +
  geom_density_ridges(alpha = 0.5,scale = 1.5) +
  labs( x = "Frequency", y = "Repeat Type") +
  theme(legend.background = element_rect(fill = NA, color = "black", linetype = "solid")) +
  coord_cartesian(clip = "off")
#dev.off()
```

```{r unmap_peak_frac_plot, eval = T, echo = F, fig.height=1.5, fig.width=3, message=F, warning = F}
peaks.plt <- peaks.frac %>%
  mutate(Region = ifelse(what == "NoWGBS","Unmapped","Mapped"))
plotpath <- file.path(plotdir,"200218_gm12878_unmapped_peak_fractions.pdf")
pdf(plotpath,height =1.5, width = 3, useDingbats = F)
ggplot(peaks.plt,aes( x = type, fill = Region, y = frac)) +
  geom_histogram(stat = "identity",position = "dodge") +
  labs( x = "Repeat Type", y = "Fraction with\nAccessibility Peak") +
  theme(legend.background = element_rect(fill = NA, color = "black", linetype = "solid")) +
  coord_cartesian(clip = "off")
dev.off()
```

#### just CGI

```{r unmap_cgi_pairwise, eval = T, echo = F, fig.height=1.5, fig.width=3, message=F, warning = F}
ggplot(meth.tb %>% filter(type == "CGI"),aes(x = gpc,y = cpg)) +
  facet_wrap(~what) +
  geom_bin2d() +
  scale_fill_gradient(trans = "log10")
```
